Scheduled banners

elarb opened the pull request:
- Makes it possible to schedule banners upfront given a date range from the admin panel
- Makes use of the newly added `/web/banners` end-point, that corresponds to [this ](https://github.com/AreaFiftyLAN/lancie-api/pull/450) PR on the API
- A new `text` tab on the admin panel has been added, which includes the newly added banner scheduling panel
- The `text` tab should contain other panels that schedule/change data on the home page (or eventually other pages) of the website, that we might want to add in the feature (https://github.com/AreaFiftyLAN/lancie-api/issues/280)
- Along the way, I separated icons used in a custom `lancie-admin` icon set, to improve performance 

Make sure to check out to [this](https://github.com/AreaFiftyLAN/lancie-api/tree/feature-scheduled_banner) branch when testing the newly added functionality.

Known issue: `update` buttons are shown everywhere after a delete. Reason for this is that a delete causes an `lancie-ajax` request and this triggers the observer for each element and thus assumes that it has been updated.

Closes https://github.com/AreaFiftyLAN/lancie-api/issues/268

elarb committed:
Add "Text" edit page and a component to be able to schedule banner text

elarb committed:
Add paper-fab

elarb committed:
Add code tag around API endpoints

elarb committed:
Use a custom icon-set

elarb committed:
Add homepage-text editor to pages

elarb committed:
Small reformatting, added custom variable for button color

elarb committed:
Remove hardcoded values

elarb committed:
Add date validation

elarb committed:
Add time-line

elarb committed:
Evenly space out components

elarb committed:
Change layout and fix update error

elarb committed:
add `paper-icon-button` and `paper-dialog`

elarb committed:
Add svg icons

elarb committed:
Add the possibility to remove all scheduled banners (with confirmation dialog); add with latest end date; sort by date

elarb committed:
Small code refactoring

TimvdLippe commented on src/lancie-admin-banner/lancie-admin-banner-scheduler.html:
```diff
@@ -0,0 +1,150 @@
+<link rel="import" href="../../bower_components/paper-input/paper-input.html">
+
+<!--Element that makes it possible to schedule banners by providing start- and enddates and a corresponding text for
+the banner within that timeframe-->
```
Documentation 👏 👏 👏

TimvdLippe commented on src/lancie-admin-content.html:
```diff
@@ -43,11 +44,12 @@
       <lancie-admin-rfid-consumption data-route="consumptions" data-path="lancie-admin-rfid/lancie-admin-rfid-consumption.html"></lancie-admin-rfid-consumption>
 
-      <lancie-admin-404 data-route="404" data-path="lancie-admin-404.html"></lancie-404>
+      <lancie-admin-404 data-route="404" data-path="lancie-admin-404.html"></lancie-admin-404>
```
🤔

martijnjanssen replied:
Yeah that one has been fixed in 4 non-merged branches now 🤔

TimvdLippe commented on src/lancie-admin-endpoint/lancie-admin-pages/lancie-admin-hptext.html:
```diff
@@ -0,0 +1,164 @@
+        --paper-fab-background: var(--lancie-button-color);
+      }
+
+      /*TODO: Should be in global scope*/
```
Hm, that's tough. Have to think about how to do that

TimvdLippe commented on src/lancie-admin-endpoint/lancie-admin-pages/lancie-admin-hptext.html:
```diff
@@ -0,0 +1,164 @@
+      }
+
+      sortOnStartDate() {
+        this.banners.sort((x, y) => x.startDate - y.startDate);
```
Does this trigger the bindings update? Else it has to be `this.banners = this.banners.sort`

TimvdLippe commented on src/lancie-admin-endpoint/lancie-admin-pages/lancie-admin-hptext.html:
```diff
@@ -0,0 +1,164 @@
+
+      getLatestDate() {
+        if (this.banners.length > 0) {
+          this.banners.sort((x, y) => x.startDate - y.startDate);
```
Why are we sorting again? And if so, it's better to call `sortOnStartDate` here

TimvdLippe commented on src/lancie-admin-icons.html:
```diff
@@ -0,0 +1,22 @@
+<link rel="import" href="../bower_components/iron-icon/iron-icon.html">
```
I like it!

TimvdLippe reviewed (commented):
Statically reviewed. Seems like a nice contribution!

DieuwerH reviewed (commented):
Error messages for the date fields are not visible / cut off

Would suggest putting the add banner button next to the remove all button

The remove all button has more space to the left than the rest of the elements

The dates in the the timeline only update when refreshing the page, not when changing the values

The flow of changing an item does not feel natural, you would expect to press the pencil to be able to change a value, not to save them. So by default all fields are disabled until the pen button is pressed. The pen image should change to a save image and upon pressing it the item is saved; disabling all the fields again and changing the button's image back to a pencil.

elarb committed:
Search for uncovered date ranges when scheduling new banner

elarb committed:
Small changes in layout

elarb commented:
@DieuwerH 

- Fixed, thanks
- Matter of preference
- I noticed as well, changed (it was `1px` btw)
- Could you show an example?
- Agreed, something I also was not sure about. Changed to an `update` button now.

martijnjanssen commented:
Can't you just fire the ajax request again to refresh the items when one is added or removed? Or at least a button to refresh the timeline display?

martijnjanssen commented:
I also recommend not placing the `lancie-admin-hptext.html` in the `lancie-admin-pages` folder, the name is not very clear for that folder, but it is more aimed at the pages with a datatable. It would be better to just make a folder for `lancie-admin-homepage-text` and put everything in there.

elarb commented:
@martijnjanssen Agreeing with the naming, will change that, but I don't really see how it doesn't fit in that folder.

elarb committed:
Rename to `lancie-admin-homepage-text`

martijnjanssen committed:
Make timeline horizontal

martijnjanssen commented:
- The `paper-dialog` is behind the drawer, this is easily fixed by using our own `lancie-dialog`.
- The circular add-button is very nice, but it doesn't really fit with the rest of our pages. Why not just use a button next to the `remove all`-button?
- I still thing we have to do something about the layout of the fields to make it more mobile or scaling friendly.
- And lastly: Nice!

elarb committed:
Replace `paper-fab` with button; make sure dialog stays on top

elarb committed:
Merge remote-tracking branch 'origin/feature-scheduled_banner' into feature-scheduled_banner

elarb commented:
`lancie-dialog` seems to behave the same: 

![chrome_2017-12-01_21-08-54](https://user-images.githubusercontent.com/18485279/33501520-262f96e4-d6dd-11e7-823c-e33187853310.png)

I fixed it though by appending the dialog to the document body after the element is being added:

![chrome_2017-12-01_21-19-47](https://user-images.githubusercontent.com/18485279/33501578-5e97b160-d6dd-11e7-8be9-9b2839e8455b.png)

elarb commented:
This is how `lancie-dialog` looks like when appending it to the document body:

![chrome_2017-12-02_18-04-12](https://user-images.githubusercontent.com/18485279/33517827-463d94a0-d78b-11e7-9dce-f286e3536e9a.png)

For now, we can either use `lancie-dialog` and live with the fact that it is behind the drawer, or we can keep it how it is (use `paper-dialog`) and fix this in a later PR (e.g. when we update `lancie-dialog`)?

elarb committed:
replace `paper-dialog` with `lancie-dialog`

elarb commented:
Alright, I think this is ready to be merged now

MatthijsKok reviewed (approved):
Love it!
